### ISSUE TITLE
continuous learning for Repairnator

### ISSUE THREAD

**Issue by monperrus** [eclipse/repairnator#970]
FYI, @javierron works on continuous learning for Repairnator, where the Travis data and Githb commits are used as data points for training Sequencer.

* [x] continuous training
* [x] continuous prediction
* [x] log parsing for prediction
* [x] debug the last problems

Next tasks after closing this one.

* context size #1033
* bigger diffs #1034
* create an artificial 6 line context for prediction #1096
* exception lines #1099
* ensemble prediction https://github.com/eclipse/repairnator/issues/1100

**Comment by monperrus** [comment]
After #971, 

- Store the scanned commit diffs that contain single-line-changes.
- Filter out repeated data or data that may be irrelevant e.g. comment lines.

**Comment by javierron** [comment]
@monperrus After re-reading the SequenceR paper, I believe storing only the .diff files that contain single line changes will not be enough. Shouldn't we store the whole class to create the abstract buggy context?

**Comment by monperrus** [comment]
You're right, SequenceR today requires the full file.

However, we can adapt Sequencer to be trained only with 3-lines unified diff.

It's a trade-off:
- full files: more call to the Github API
- 3-lines unified default diff: one single call

It would be great to see how better is the training with the full context.

**Comment by monperrus** [comment]
Per our discussion today with @javierron, For continuous learning, the TensorFlow implementation of OpenNMT would be good, but it does not support the copy mechanism.

**Comment by javierron** [comment]
> Per our discussion today with @javierron, For continuous learning, the TensorFlow implementation of OpenNMT would be good, but it does not support the copy mechanism.

Issue is documented in https://github.com/OpenNMT/OpenNMT-tf/issues/95

**Comment by monperrus** [comment]
Related work on vocabulary adaptation: On-the-Fly Adaptation of Source Code Models using Meta-Learning
https://arxiv.org/pdf/2003.11768.pdf

**Comment by monperrus** [comment]
Per our discussion, we can do greedy optimization on any benchmark metric, incl. D4J/BEARS compilation rate and D4J/BEARS plausible rate.

**Comment by javierron** [comment]
OpenNMT ensemble translation does not currently work.
Issue: https://github.com/OpenNMT/OpenNMT-py/issues/1252

**Comment by javierron** [comment]
Continual learning research:
https://github.com/prprbr/awesome-lifelong-continual-learning

**Comment by monperrus** [comment]
OK, then moved task to top-level issue https://github.com/eclipse/repairnator/issues/1100

What about ticking now the last todo and closing this issue?

**Comment by javierron** [comment]
> What about ticking now the last todo and closing this issue?

The RepairnatorZero loop is running now, I think we can close this issue.

**Comment by monperrus** [comment]
Then hurray we close this one :champagne: thanks for the hard work.

**Comment by monperrus** [comment]
For the record, if Sequencer has more than one patch, we only create a single PR, with the patch with highest likelihood from OpenNMT

**Comment by monperrus** [comment]
For the record, the problem we face is called "catastrophic forgetting"


> Early attempts to mitigate catastrophic forgetting typically consisted of memory systems that store previous data and that regularly replay old samples interleaved with samples drawn from the newdata  (Robins  1993,  1995),  and  these  methods  are  still  used  today  (Gepperth  &  Karaoguz  2015, Rebuffi et al. 2016).
page 8

Continual Lifelong Learning with Neural Networks:A Review
https://arxiv.org/pdf/1802.07569.pdf

Advanced techniques:

* EWC from Overcoming catastrophic forgetting in neural networks https://arxiv.org/pdf/1612.00796.pdf
* Progress & compress: A scalable framework for continual learning. InICML, 2018 https://arxiv.org/pdf/1805.06370
* CLEAR: Experience Replay for Continual Learning 2019 http://papers.nips.cc/paper/8327-experience-replay-for-continual-learning.pdf

**Comment by monperrus** [comment]
FYI:
Learning to Prompt for Continual Learning (L2P)  by  Google Research
https://github.com/google-research/l2p
